### ISSUE TITLE
Better textDocument/hover support for property-like descriptors

### ISSUE THREAD

**Issue by Rapptz** [microsoft/pyright#3291]
**Is your feature request related to a problem? Please describe.**

Hover and documentation tooltips incorrectly mark non-callable descriptors as `(method)` which causes visible user confusion in my support channels due to users thinking they must be called as functions, rather than being marked as `(property)` or `(descriptor)` to guide the user to not call it as a function. Likewise, custom descriptors do not get documentation filled in.

**Describe the solution you'd like**

It would be ideal if custom descriptors that behaved like properties had a label signaling users to not call it as a method. Something like `(property)` or `(descriptor)` works okay for me. It would also be ideal if the hover would show documentation for these property-like descriptors.

**Additional context**

I'm the maintainer of a popular library named [discord.py](https://github.com/Rapptz/discord.py) with hundreds of thousands of users. Many of these users use Pylance or Pyright as their type checker due to it being promoted by me as the recommended type checker of choice due to its breadth of features and quality. My library is complicated in order to provide users with a good developer experience and some of this complication involves a lot of custom descriptors that behave like properties but aren't `property` itself.

Examples of these properties are things like `cached_property`, `cached_slot_property`, and `flag_value`. Over time, Pyright's implementation of custom descriptors have improved and these descriptors work fine in terms of actual type checking, though the LSP capabilities still kind of hardcode these as not-property. This does not only affect my library, but the standard library as well. For example, with the built-in `functools.cached_property` decorator, the `textDocument/hover` request does not show `(property)` or the docstring:

```py
from functools import cached_property

class A:
    @cached_property
    def foo(self) -> int:
        """This is foo docstring"""
        return 10

a = A()
a.foo  # hover
```

![image](https://user-images.githubusercontent.com/1695103/161374083-ec4b85f6-de99-435e-bed1-2ad75638e383.png)

If a `property` is used instead, this works as expected:

![image](https://user-images.githubusercontent.com/1695103/161374108-bbf36da1-f977-4cbc-a67f-64c6eda6527a.png)

While it would be possible for pyright to special case `functools.cached_property` much like it special cases `property`, I think this would be a bad hit for library users who are making use of custom descriptors that operate like properties. There is a way to "cheat" to Pyright by marking the descriptor as a `NewType`:

```py
from typing import TYPE_CHECKING, NewType

if TYPE_CHECKING:
    cached_property = NewType('cached_property', property)
else:
    from functools import cached_property
```

However, this does not work for more sophisticated descriptors such as a `cached_slot_property` that mentions which slot to cache the property result to, such as:

```py
class B:
    __slots__ = ('foo')

    @cached_slot_property('foo')
    def bar(self) -> int:
        return 10
```

The implementation of this descriptor cannot be "cheated" using `NewType`.

**Possible Contribution**

I wasn't sure where to put this, but I was willing to make a pull request to fix this though according to the guidelines I thought this might require making an issue first.

**Comment by erictraut** [comment]
Thanks for the suggestion. I agree that displaying a descriptor as a "method" is confusing.

Do you think that most Python developers are familiar enough with the term "descriptor" to understand what it means? Would it be preferable to use the term "property" to refer to all descriptors, or would that be confusing?

Presumably this should apply to any object that has a `__get__` method, even if it doesn't have a corresponding `__set__` method?

**Comment by Rapptz** [comment]
Thank you for the prompt response.

> Do you think that most Python developers are familiar enough with the term "descriptor" to understand what it means? Would it be preferable to use the term "property" to refer to all descriptors, or would that be confusing?

I agree that `property` is probably better than `descriptor` for the average user. `descriptor` is more of a deeper knowledge of Python kind of detail that I feel novices wouldn't quite understand.

> Presumably this should apply to any object that has a __get__ method, even if it doesn't have a corresponding __set__ method?

Yeah, I think that's for the best.

**Comment by erictraut** [comment]
I don't think it's feasible to support docstrings for general-purpose descriptors (or for `cached_property`) because there's no way for the type checker to associate the decorated method with the `__get__` call.

I have added support for propagating docstrings for ParamSpecs, but that doesn't apply in this case.

Obviously, `property` has a bunch of special-case logic that includes the propagation of docstrings, but I don't think this can be done for generalized descriptors. Let me know if you have any ideas here.

**Comment by erictraut** [comment]
Here's a potential heuristic:

If the following are true:
1. A class contains a `__get__` method
2. The `__get__` method doesn't contain a docstring
3. The class has an `__init__` method that accepts a callable as its first parameter

If all of these are true, perhaps it makes sense for the `__get__` method to "adopt" the docstring from the callable that is passed to the class constructor.

I'll play around with this and see if it works in the majority of cases.

**Comment by erictraut** [comment]
@Rapptz, I noticed that your library isn't "py.typed" and doesn't contain any inlined type annotations. Would you consider making that investment? We're encouraging all library authors to do so because it greatly improves the experience for consumers of the library — both in terms of language server features and static type checking. We've written [this guidance](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) to help library maintainers who are interested in adding inlined type information.

**Comment by Rapptz** [comment]
I think for general-purpose descriptors using the class's function declaration's docstring like is done with regular properties would be fine, though I'm unsure if there's a degenerative case where this would be undesirable.

When I was hacking around to see the feasibility, I found through the debugger that it goes through this code for the remaining cases: 

https://github.com/microsoft/pyright/blob/adaab317f5b7a03fdd3bbee1d068c7e8347a80b0/packages/pyright-internal/src/languageService/tooltipUtils.ts#L123-L129

And within the calls it fails the `isProperty` check. 

> Here's a potential heuristic:

I'm not sure if my classes personally meet the 3rd check, I know for example my `cached_slot_property` has the `name` as the first parameter and not the second one. I could obviously change it if it makes things easier since it's a private type but I'm unsure about others.

>  I noticed that your library isn't "py.typed" and doesn't contain any inlined type annotations. 

The PyPI version of the library is untyped since it's v1.7.3, for those users we recommend using a separate stubs library. However on GitHub and the pre-release version we do have things typed inline and have a py.typed file with 100% verifytypes coverage.

**Comment by erictraut** [comment]
> However on GitHub and the pre-release version we do have things typed inline and have a py.typed file with 100% verifytypes coverage.

You just put a big smile on my face. :)

**Comment by erictraut** [comment]
I've updated the hover provider so it now displays `(property)` for any field that includes a descriptor object (or union of descriptor objects). And I've updated the logic used by the hover and completion providers to propagate a docstring from a method that is decorated with a generalized descriptor class. I softened the heuristic to remove the "first parameter" requirement, so it should work more generally.

**Comment by Rapptz** [comment]
I've tested the changes locally and it seems to work great, thanks a lot!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.235, which I just pblished. It will also be included in the next release of pylance.
